### PR TITLE
Refactor location module

### LINKED ISSUE #2772: Location module has a confusing interface
## Type

- Enhancement

## Problem description

The edit page of a location is confusing as it has 2 forms, 2 save buttons and they both behave differently.

## Steps to reproduce

https://demo.fork-cms.com/private/en/location/edit?token=z4pjcflp2s&id=1

- Edit something in both forms, click "update map" (whatever that means). The top form will be submitted and handled but the bottom form will be ignored.
- Edit something in both forms, click "save". The bottom form will be submitted using AJAX and after succesful completion it will submit the top form which makes no sense whatsoever.

## Expected behavior / Proposed solutions

I expect the form to make sense. There is no purpose (as far as I can tell) for the AJAX call or even the separate forms. 

My suggestion:

- Split up the form in 2 tabs, address and map settings
- Lose the AJAX call, there should be 1 "save" button that saves the entire form
- Add a checkbox to the "map settings" tab to indicate whether to overwrite the general map settings set on the Index action.
- Disable/enable the map settings fields based on the checkbox value.

### PR DESCRIPTION
**Pull request by bjorvack**
## Type

- Critical bugfix

## Resolves the following issues

fixes #2772, #3101

## Pull request description

Refactor Location module so no Ajax is used and only has one save button

### CODE REVIEW

File: src/Backend/Modules/Location/Actions/Edit.php [hunk 1]
```diff
@@ -110,10 +108,8 @@ private function loadForm(): void
         $this->form->addText('city', $this->record['city'])->makeRequired();
         $this->form->addDropdown('country', Intl::getRegionBundle()->getCountryNames(BL::getInterfaceLanguage()), $this->record['country'])->makeRequired();
         $this->form->addHidden('redirect', 'overview');
-    }
 
-    protected function loadSettingsForm(): void
-    {
+
```
Comment (Katrienvh) [inline src/Backend/Modules/Location/Actions/Edit.php:112]: I think there are 2 newlines, only need one right?

File: src/Backend/Modules/Location/Js/Location.js [hunk 2]
```diff
@@ -19,29 +19,29 @@ export class Location {
     this.width = null
     this.zoomLevel = null
 
+    $('[data-role=toggle-settings]').on('change', function () {
+      console.log('here', $(this))
```
Comment (Katrienvh) [inline src/Backend/Modules/Location/Js/Location.js:23]: forgot to remove log

File: src/Backend/Modules/Location/Js/Location.js [hunk 3]
```diff
@@ -19,29 +19,29 @@ export class Location {
     this.width = null
     this.zoomLevel = null
 
+    $('[data-role=toggle-settings]').on('change', function () {
```
Comment (Katrienvh) [inline src/Backend/Modules/Location/Js/Location.js:22]: use arrow function

File: src/Backend/Modules/Location/Js/Location.js [hunk 4]
```diff
@@ -141,38 +141,10 @@ export class Location {
...
-          }
-
-          $('input#redirect').val('edit')
-          $('form#edit').submit()
-        }
-      }
-    })
+  setCenter () {
+    this.getMapData()
+    $('#centerLat').val(this.centerLat)
```
Comment (Katrienvh) [inline src/Backend/Modules/Location/Js/Location.js:146]: please use data attributes as selector in the JS.

**Review by Katrienvh — CHANGES_REQUESTED** [review]

File: src/Backend/Modules/Location/Js/Location.js (hunk 3 above)
Comment (bjorvack) [reply src/Backend/Modules/Location/Js/Location.js:22]: fixed f8da312732f8d3a6276e8f89da6236c219e13433

File: src/Backend/Modules/Location/Js/Location.js (hunk 2 above)
Comment (bjorvack) [reply src/Backend/Modules/Location/Js/Location.js]: fixed f8da312732f8d3a6276e8f89da6236c219e13433

File: src/Backend/Modules/Location/Actions/Edit.php (hunk 1 above)
Comment (bjorvack) [reply src/Backend/Modules/Location/Actions/Edit.php]: fixed 213c93c02cbf040c635e7011ff04fb8d619dfab2

File: src/Backend/Modules/Location/Js/Location.js (hunk 3 above)
Comment (bjorvack) [reply src/Backend/Modules/Location/Js/Location.js]: fixed bfb7b8e6ab81b58d0a824405378afac64cf475c0

File: src/Backend/Modules/Location/Js/Location.js (hunk 4 above)
Comment (bjorvack) [reply src/Backend/Modules/Location/Js/Location.js:145]: fixed bfb7b8e6ab81b58d0a824405378afac64cf475c0

**Review by carakas — CHANGES_REQUESTED** [review]
Phpstan fails

File: src/Backend/Modules/Location/Actions/Add.php [hunk 5]
```diff
@@ -74,6 +74,17 @@ private function validateForm(): void
...
+                $generalSettings = $this->get('fork.settings')->getForModule('Location');
+                $center = ['lat' => $item['lat'], 'lng' => $item['lng']];
+                BackendLocationModel::setMapSetting($item['id'], 'zoom_level', (string) $generalSettings['zoom_level']);
+                BackendLocationModel::setMapSetting($item['id'], 'map_type', (string) $generalSettings['map_type']);
+                BackendLocationModel::setMapSetting($item['id'], 'map_style', (string) 'standard');
+                BackendLocationModel::setMapSetting($item['id'], 'center', $center);
+                BackendLocationModel::setMapSetting($item['id'], 'height', (int)  $generalSettings['height']);
+                BackendLocationModel::setMapSetting($item['id'], 'width', (int)  $generalSettings['width']);
+                BackendLocationModel::setMapSetting($item['id'], 'directions', false);
+                BackendLocationModel::setMapSetting($item['id'], 'full_url', false);
```
Comment (carakas) [inline src/Backend/Modules/Location/Actions/Add.php:86]: not a huge fan of how this looks but you don't need to change this

File: src/Backend/Modules/Location/Actions/Add.php [hunk 6]
```diff
@@ -74,6 +74,17 @@ private function validateForm(): void
                 // insert the item
                 $item['id'] = BackendLocationModel::insert($item);
 
+                $generalSettings = $this->get('fork.settings')->getForModule('Location');
```
Comment (carakas) [inline src/Backend/Modules/Location/Actions/Add.php:77]: You need to write extra code that when these settings change you update all locations that don't overwrite the settings

**Review by carakas — CHANGES_REQUESTED** [review]

File: src/Backend/Modules/Location/Actions/Add.php (hunk 6 above)
Comment (bjorvack) [reply src/Backend/Modules/Location/Actions/Add.php:77]: added update script on settings page 03dae81bd594410725572bee8e1aad4459176665

**Review by carakas — APPROVED** [review]

File: src/Backend/Modules/Location/Js/Location.js [hunk 7]
```diff
@@ -70,7 +69,7 @@ export class Location {
       // add event listener
       google.maps.event.addListener(marker, 'dragend', function () {
```
Comment (Katrienvh) [inline src/Backend/Modules/Location/Js/Location.js:70]: do you need the `.bind` on line 72 when you use an arrow function?

Comment (bjorvack) [reply src/Backend/Modules/Location/Js/Location.js]: @Katrienvh fixed in fec96fedb131dbc00520ea861339093e2eefe5c4

File: src/Backend/Modules/Location/Js/Location.js (hunk 2 above)
Comment (Katrienvh) [inline src/Backend/Modules/Location/Js/Location.js]: Console.log weghalen?

**Review by Katrienvh — CHANGES_REQUESTED** [review]

File: src/Backend/Modules/Location/Js/Location.js (hunk 2 above)
Comment (Katrienvh) [reply src/Backend/Modules/Location/Js/Location.js]: wait this is outdated code? Strange it was still in my review

Comment (Katrienvh) [reply src/Backend/Modules/Location/Js/Location.js]: Ignore this

**Review by Katrienvh — APPROVED** [review]

**Pull request merged**